### PR TITLE
zookeeper: parse server responses

### PR DESCRIPTION
**Pull request by rgs1**
This adds support for parsing server responses and watch events. One
or more follow-up PRs will add support for latency measurements and
access log.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
Note: docs & tests still missing, will follow-up later today/tomorrow.

**Comment by rgs1** [comment]
@mattklein123 getting this twice in a row now:

```
//test/extensions/filters/http/dynamic_forward_proxy:proxy_filter_integration_test TIMEOUT in 315.0s
  /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-dbg/testlogs/test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_integration_test/test.log
```

**Comment by rgs1** [comment]
I'll push tests today/tomorrow.

**Comment by rgs1** [comment]
hmm, unrelated failure:

```
//test/common/runtime:runtime_impl_test                                 TIMEOUT in 315.0s
  /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/runtime/runtime_impl_test/test.log
```

**Comment by rgs1** [comment]
another timeout for the same unrelated integration test:

```
//test/common/runtime:runtime_impl_test                                 TIMEOUT in 315.0s
  /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/runtime/runtime_impl_test/test.log
```
cc: @alyssawilk since maybe this is related to e61681d2268 ?

**Comment by rgs1** [comment]
Yeah it does time out in the test case added by e61681d:

```
[----------] 3 tests from StaticLoaderImplTest
[ RUN      ] StaticLoaderImplTest.All
[       OK ] StaticLoaderImplTest.All (95 ms)
[ RUN      ] StaticLoaderImplTest.ProtoParsing
[       OK ] StaticLoaderImplTest.ProtoParsing (147 ms)
[ RUN      ] StaticLoaderImplTest.RuntimeFromNonWorkerThreads
Terminated
-- Test timed out at 2019-07-28 19:03:52 UTC --
```

**Comment by rgs1** [comment]
Possible fix for timeouts: #7746.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 1]
```diff
@@ -397,20 +447,71 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
...
+  ensureMinLength(len, PROTO_VERSION_LENGTH + TIMEOUT_LENGTH + SESSION_LENGTH + INT_LENGTH);
+
+  auto timeout = helper_.peekInt32(data, offset);
+
+  // Skip session id + password.
+  offset += SESSION_LENGTH;
+  skipString(data, offset);
+
+  // Read readonly flag, if it's there.
+  bool readonly{};
```
Comment (snowp) [inline source/extensions/filters/network/zookeeper_proxy/decoder.cc]: should this bit be a helper? `maybeReadBool` or something?

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 2]
```diff
@@ -156,8 +152,62 @@ void DecoderImpl::decode(Buffer::Instance& data, uint64_t& offset) {
...
+  }
+
+  // Find the corresponding request for this XID.
+  const auto it = requests_by_xid_.find(xid);
+  if (it == requests_by_xid_.end()) {
+    return;
   }
+
+  const auto opcode = it->second;
+  requests_by_xid_.erase(it);
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc]: those don't get added.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 3]
```diff
@@ -156,8 +152,62 @@ void DecoderImpl::decode(Buffer::Instance& data, uint64_t& offset) {
...
+  case XidCodes::WATCH_XID:
+    parseWatchEvent(data, offset, len, zxid, error);
+    return;
+  default:
+    break;
+  }
+
+  // Find the corresponding request for this XID.
+  const auto it = requests_by_xid_.find(xid);
+  if (it == requests_by_xid_.end()) {
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc]: Should never happen actually. Yeah we could assert or log.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 4]
```diff
@@ -397,20 +447,71 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
...
+  if (data.length() >= offset + 1) {
+    readonly = helper_.peekBool(data, offset);
+  }
+
+  callbacks_.onConnectResponse(0, timeout, readonly);
+}
+
+void DecoderImpl::parseWatchEvent(Buffer::Instance& data, uint64_t& offset, const uint32_t len,
+                                  const int64_t zxid, const int32_t error) {
+  ensureMinLength(len, SERVER_HEADER_LENGTH + (3 * INT_LENGTH));
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc]: 1 string = 1 int (length) + chars

the string could be empty, so the min is 2 ints + 1 int (the str length).

File: source/extensions/filters/network/zookeeper_proxy/filter.cc [hunk 5]
```diff
@@ -236,6 +234,138 @@ void ZooKeeperFilter::onCloseRequest() {
...
+  case OpCodes::GETCHILDREN:
+    opname = "getchildren_resp";
+    config_->stats_.getchildren_resp_.inc();
+    break;
+  case OpCodes::GETCHILDREN2:
+    opname = "getchildren2_resp";
+    config_->stats_.getchildren2_resp_.inc();
+    break;
+  case OpCodes::DELETE:
+    opname = "remove_resp";
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/filter.cc]: yeah it is, libraries are inconsistent about this too. I'll rename it.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 6]
```diff
@@ -397,20 +447,71 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
...
+      //       internal cursor is reset for each individual message.
+      helper_.reset();
+
       const uint64_t current = offset;
-      decode(data, offset);
-      callbacks_.onRequestBytes(offset - current);
+      switch (dtype) {
+      case DecodeType::READ:
+        decodeOnData(data, offset);
+        callbacks_.onRequestBytes(offset - current);
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/extensions/filters/network/zookeeper_proxy/decoder.cc]: I think I'll leave it there for now, in a follow-up diff i'll add access logging support and we'd want that to happen after each successful decode call.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 7]
```diff
@@ -156,8 +152,64 @@ void DecoderImpl::decode(Buffer::Instance& data, uint64_t& offset) {
...
+    return;
+  default:
+    break;
+  }
+
+  // Find the corresponding request for this XID.
+  const auto it = requests_by_xid_.find(xid);
+  if (it == requests_by_xid_.end()) {
+    // If this happens, it's a server-side bug.
+    ASSERT(false);
```
Comment (snowp) [inline source/extensions/filters/network/zookeeper_proxy/decoder.cc]: how about doing `ASSERT(it != requests_by_xid_.end())`?

**Pull request merged**